lib/sharepoint/client.rb

```diff
@@ -6,23 +6,45 @@
 
 require 'active_support/core_ext/string/inflections'
 require 'active_support/core_ext/object/blank'
+require 'sharepoint/client/token'
 
 module Sharepoint
   class Client
     FILENAME_INVALID_CHARS = '~"#%&*:<>?/\{|}'
 
+    def authenticating_with_token
+      generate_new_token
+      yield
+    end
+
+    def generate_new_token
+      token.retrieve
+    end
+
+    def bearer_auth
+      "Bearer #{token}"
+    end
+
     # @return [OpenStruct] The current configuration.
     attr_reader :config
+    attr_reader :token
 
     # Initializes a new client with given options.
     #
     # @param [Hash] config The client options:
     #  - `:uri` The SharePoint server's root url
+    #  - `:authentication` The authentication method to use [:ntlm, :token]
     #  - `:username` self-explanatory
     #  - `:password` self-explanatory
+    #  - `:client_id` self-explanatory
+    #  - `:client_secret` self-explanatory
+    #  - `:tenant_id` self-explanatory
+    #  - `:cert_name` self-explanatory
+    #  - `:auth_scope` self-explanatory
```

Comment:
We can probably use a `Struct` here, because with openstruct you can pass anything, and `ethon_easy_options` is not even listed.

But let's leave this for a future refactor

Reply:
https://github.com/ifad/sharepoint/issues/41